pkg/cpe/wfn.go

```diff
@@ -0,0 +1,209 @@
+		qRun, atStart = false, true
+	)
+	for i, r := range s {
+		switch r {
+		case '*':
+			if esc { // if escaped, it's not special
+				break
+			}
+			if i != 0 && i != last {
+				return fmt.Errorf("cpe: invalid position for special character: %q at %d", r, i)
+			}
+		case '?':
+			if esc { // if escaped, it's not special
+				break
+			}
+			qRun = true
+		case '\\':
+			esc = true
+			// skip resetting the esc bool
+			continue
+		default:
+			// The only place valid for a run of '?' is at the beginning and end
+			// of a string. So if we read a series of '?' then a non-'?'
+			// *twice*, that's invalid.
+			if qRun && !atStart {
+				return fmt.Errorf("cpe: invalid position for special character: %q at %d", '?', i-1)
+			}
+			if reserved(r) && !esc {
+				return fmt.Errorf("invalid unquoted character: %q at %d", r, i)
+			}
+		}
+		if r != '?' {
+			qRun, atStart = false, false
+		}
+		esc = false
+	}
+	return nil
+}
+
+// NewValue constructs a specific value and ensures it's a valid string.
+//
+// This function does not quote the provided string, only validates that the
+// quoting is proper.
+func NewValue(v string) (Value, error) {
+	if err := validate(v); err != nil {
+		return Value{}, err
+	}
+	return Value{
+		Kind: ValueSet,
+		V:    v,
+	}, nil
+}
+
+// Value represents all the states for an attribute's value.
+type Value struct {
+	Kind ValueKind
+	V    string
+}
+
+// ValueKind indicates what "kind" a value is.
+type ValueKind uint
+
+//go:generate stringer -type ValueKind
+
+// These are the valid states for a wfn attribute's value.
+const (
+	ValueUnset ValueKind = iota
+	ValueAny
+	ValueNA
+	ValueSet
+)
+
+func (v *Value) String() string {
+	var b strings.Builder
+	v.bind(&b)
+	return b.String()
+}
+
+// WFN is a well-formed name as defined by the Common Platform Enumeration (CPE)
+// spec: https://nvlpubs.nist.gov/nistpubs/Legacy/IR/nistir7695.pdf
+//
+// This package does not implement binding into URI form.
+type WFN struct {
+	Attr [NumAttr]Value
+}
+
+// Valid reports an error if a WFN is not, in fact, well-formed.
+//
+// Functions returning a WFN call this before returning, but if a WFN is
+// constructed in code, the user should check it via this method.
+func (w WFN) Valid() error {
+	unset := 0
+	for i := 0; i < NumAttr; i++ {
+		if err := validate(w.Attr[i].V); err != nil {
+			return fmt.Errorf("cpe: wfn attr %v is invalid: %w", Attribute(i), err)
+		}
+		if v := &w.Attr[i]; v.Kind == ValueUnset {
+			unset++
+		}
+	}
+	if unset == NumAttr {
+		return ErrUnset
+	}
+	if p := w.Attr[int(Part)]; p.Kind == ValueSet {
+		if len(p.V) != 1 ||
+			(p.V != "a" && p.V != "o" && p.V != "h") {
+			return fmt.Errorf("cpe: wfn attr %v is invalid: %q is a disallowed value", Part, p.V)
+		}
+	}
+	return nil
+}
+
+// ErrUnset is returned from (WFN).Valid() if it is the zero value.
+var ErrUnset = errors.New("cpe: wfn is empty")
+
+func (w WFN) String() string {
+	return w.BindFS()
```

Comment:
This is a little strange and also makes a computation happen on every call to String().

Would it make sense to create a WFN constructor, validate the WFN, Bind it, and set a Repr attribute all in one go? Then you no longer need to cll this w.BindFS() all the time, and you also don't need conflate whether a user should use String() or BindFS() method.

Reply:
The string has to be created on every call because the structure isn't immutable. To make it so the string is only created once, the values would have to be made private and everything mediated by methods that keep the string up to date, instead of just creating the string on demand.

Having the WFN implement Stringer is just a nicety and we can remove this method if you think it'll be abused.

Reply:
Should a constructed WFN be immutable?